### ISSUE TITLE
[meta] Verify the RALF public data path: placeholder retrieval features and divergent provenance

### ISSUE THREAD

**Issue by creative-graphic-design-dev** [creative-graphic-design/design-generators#303]
Two observations suggest the public RALF data path may not carry what the method needs, and that parity may not cover that path.

Placeholder retrieval features. `models/ralf/src/ralf/datasets.py:263-271` returns retrieved candidates whose image and saliency tensors are zeros with 1x1 spatial dimensions:

```python
batch, candidates = indexes.shape
return RalfRetrievedBatch(
    image=torch.zeros(batch, candidates, 3, 1, 1),
    saliency=torch.zeros(batch, candidates, 1, 1, 1),
    ...
)
```

RALF conditions generation on retrieved layouts together with their visual features, so constant zeros would remove that signal if the model consumes these fields.

Divergent data provenance. The original implementation reads the author-distributed `cache.zip` bundle, which contains a preprocessed CGL dataset, the authors' splits, and precomputed DreamSim retrieval correspondences (`vendor/ralf/README.md:109-127`). The package instead loads `creative-graphic-design/CGL-Dataset` with `name="ralf-style"` from the Hub. Unless that config reproduces the authors' preprocessing and splits, the two paths do not see the same data.

Investigate and report before any code change:

- Whether the model actually reads `image` and `saliency` from the retrieved batch, and if so what the zeros change; whether the placeholder is a deliberate, documented limitation.
- Whether `ralf-style` reproduces the authors' preprocessing, splits, and retrieval correspondences.
- Which data path the parity suite exercises: the vendor cache, the Hub path, or both, and therefore whether the public path is verified at all.

This gates the S5 training reproduction in #270, which builds on the same data path, so resolve it before that work continues.

Requested by the user 2026-08-20.

**Comment by creative-graphic-design-dev** [comment]
## Retrieved images are unused in every current configuration, and the Hugging Face Hub dataset is still unverified

RALF, the retrieval-augmented layout model this package ports, generates a layout for a background image by conditioning on layouts retrieved from the training set. This issue raised three questions about that retrieval input. Does the model actually read the retrieved images and their saliency maps, the single-channel maps whose values say how visually prominent each pixel of an image is, which the package fills with zeros? Does the dataset published on the Hugging Face Hub reproduce the preprocessing, the train and validation splits, and the retrieval correspondences that the authors distributed in their own data bundle? And which of those two data sources do the comparison checks actually run on, the checks that feed the same input to this package and to the authors' original code and compare the results?

The code answers the first and third questions. The second is still unanswered, because it needs a dataset-to-dataset comparison that nobody has run.

One concern is resolved and one remains. Retraining this package locally to confirm that its training code behaves like the authors' code is unaffected, because that work reads the authors' data bundle and not the Hugging Face Hub dataset. What still needs checking is the path a user takes when generating layouts with the published package, which does read the Hugging Face Hub dataset and the zero-filled retrieval fields.

### Does the model read the zero-filled retrieved images?

No. Each retrieved candidate reaches the model as layout geometry only: the center coordinates, width, height, and category label of each element, plus a mask marking which elements are real. Those numbers go through a layout feature extractor taken from FIDNet, a network the RALF authors trained to score layouts and whose encoder RALF reuses to turn one whole layout into a single vector. The retrieved images and saliency maps are shaped and carried alongside, and then no encoder reads them.

The original code encodes retrieved images only when its `use_reference_image` switch is on, and it appends those image features to the layout features in the same place. That switch is off by default in both implementations, and no converted model file or training configuration turns it on. For every configuration in the repository today, the zeros therefore change nothing, because neither implementation looks at retrieved pixels.

### What happens if reference-image encoding is ever switched on?

The package would quietly become a different model. It reads the switch only to size the position numbering that tells the model where each item sits in its input sequence, and it has no code path that encodes retrieved images at all. A released model file that turns the switch on would lose the image half of the combined input that the original code appends, and it would do so without raising an error. No converted model file turns it on, so nothing is affected today. Before one that does is published, the package needs either the missing encoding branch or an explicit failure on that configuration.

### Which data source do the comparison checks run on?

The authors' own bundle, on both sides. The zero-filled placeholder lives in the generation helper, not in the training path. The training loader builds each retrieved candidate from real pixels, reading that candidate's image and saliency map straight from the dataset row.

The dataset it reads is not the Hugging Face Hub one. It reads the split files of the authors' bundle together with the retrieval table the authors precomputed with DreamSim, an image-similarity model that decides which training images are the nearest neighbours of a given input. Both implementations read that same bundle throughout the comparison checks, including the longest one, which trained the package and the original code side by side through ninety consecutive weight updates, each computed from a group of thirty-two samples, and which drew its retrieval candidates from the bundle's own train and validation tables.

The loader comparison then ran both loaders over that bundle and found them identical. Each produced the same 48,544 training and 6,002 validation samples, identified by the dataset's own sample identifiers, with no sample appearing in both splits. Reducing each side's full ordered stream of numbers, layout values and retrieval values together, to a single SHA-256 fingerprint, a short code that changes whenever any byte of the input changes, produced the same fingerprint on both sides, so the streams are identical rather than merely similar. The authors' data path is therefore verified, including the retrieved pixels that the placeholder would have zeroed.

One caveat belongs on the record. The loader's image helper substitutes an all-zero image when a row carries no usable image, so a bundle missing saliency maps would be zeroed with no error. The loader comparison rules that out for the data actually used, but the substitution is silent.

### What remains open

Nothing has tested the Hugging Face Hub dataset. No recorded comparison touches `creative-graphic-design/CGL-Dataset` with `name="ralf-style"`, so the second question stands: whether that configuration reproduces the authors' preprocessing, splits, and retrieval correspondences is unknown, and so is the generation path that reads the placeholder.

That splits the work in two:

- Verifying the public path, which stays open here: compare the Hugging Face Hub dataset against the authors' splits and preprocessing, then decide whether the placeholder in the generation path is a documented limitation or a defect.
- Reproducing training locally, which can proceed: it consumes the authors' data path that the loader comparison already verified, so neither observation here constrains it.

The local training reproduction work is [pull request #270](https://github.com/creative-graphic-design/design-generators/pull/270).

<details>
<summary>Where to read each claim in the source</summary>

Each link opens the exact lines that establish the claim. The original implementation links point at the upstream RALF repository at the revision this repository pins.

| Claim | Source |
| --- | --- |
| Retrieved candidates enter as layout geometry only | [`modeling_ralf.py` lines 865-893](https://github.com/creative-graphic-design/design-generators/blob/bd14248/models/ralf/src/ralf/modeling_ralf.py#L865-L893) |
| Retrieved images and saliency are carried but unread | [`modeling_ralf.py` lines 1208-1211](https://github.com/creative-graphic-design/design-generators/blob/bd14248/models/ralf/src/ralf/modeling_ralf.py#L1208-L1211) |
| The reference-image switch is off by default | [`configuration_ralf.py` line 119](https://github.com/creative-graphic-design/design-generators/blob/bd14248/models/ralf/src/ralf/configuration_ralf.py#L119) |
| The package reads the switch only for position numbering | [`modeling_ralf.py` lines 998 and 1050](https://github.com/creative-graphic-design/design-generators/blob/bd14248/models/ralf/src/ralf/modeling_ralf.py#L998-L1050) |
| The original encodes and appends reference images under the switch | [`retrieval_augmented_autoreg.py` lines 542-576](https://github.com/CyberAgentAILab/RALF/blob/c51db6032acbd0bd0ce72433becce08317e7874d/image2layout/train/models/retrieval_augmented_autoreg.py#L542-L576) |
| The zero-filled placeholder in the generation helper | [`datasets.py` lines 263-271](https://github.com/creative-graphic-design/design-generators/blob/bd14248/models/ralf/src/ralf/datasets.py#L263-L271) |
| The training loader stacks real retrieved pixels | [`datamodule.py` lines 177-205](https://github.com/creative-graphic-design/design-generators/blob/93085ea/models/ralf/src/ralf/training/datamodule.py#L177-L205) |
| The training loader reads the authors' bundle and retrieval table | [`datamodule.py` lines 413-450](https://github.com/creative-graphic-design/design-generators/blob/93085ea/models/ralf/src/ralf/training/datamodule.py#L413-L450) |
| The silent all-zero image substitution | [`datamodule.py` line 79](https://github.com/creative-graphic-design/design-generators/blob/93085ea/models/ralf/src/ralf/training/datamodule.py#L79) |

</details>
